polyfills/systemxhr/systemxhr.js

```diff
@@ -0,0 +1,430 @@
+        set: function(cb) {
+          this._onchange(changeEvent, cb);
+        }
+      });
+    });
+
+    this.serialize = function() {
+      return {
+        id: reqId,
+        data: {
+          operation: 'createXMLHttpRequest',
+          options: extraData.options
+        },
+        processAnswer: function(answer) {
+          if (!answer.error) {
+            _resolve(answer.id);
+          } else {
+            _reject(answer.error);
+          }
+        }
+      };
+    };
+
+    this._onreadystatechange();
+  }
+
+  function FakeXMLHttpRequestUpload() {
+    // Inherited properties from XMLHttpRequestEventTarget
+    var onChangeEvents = [
+      'abort',
+      'error',
+      'load',
+      'loadend',
+      'loadstart',
+      'progress',
+      'timeout',
+    ];
+
+    onChangeEvents.forEach(changeEvent => {
+      Object.defineProperty(this, 'on' + changeEvent, {
+        set: function(cb) {
+          this._onchange(changeEvent, cb);
+        }
+      });
+    });
+  }
+
+  var navConnPromise = new NavConnectHelper(SYSTEMXHR_SERVICE);
+  var realXMLHttpRequest = window.XMLHttpRequest.bind(window);
+
+  function XMLHttpRequestShim(options) {
+    if (options && options.mozSystem) {
+      return navConnPromise.createAndQueueRequest({
+                                                    options: options
+                                                  }, FakeXMLHttpRequest);
+    } else {
+      return new realXMLHttpRequest(options);
+    }
+  };
+
+  window.XMLHttpRequest = XMLHttpRequestShim;
+
+  navConnPromise.then(function() {}, e => {
+    debug('Got an exception while connecting ' + e);
+    window.XMLHttpRequest = realXMLHttpRequest;
+  });
+
+  function FakeEventTarget(listenerCb, field, promise) {
+    // _listeners[type][ListenerId] => undefined or a callback function
+    var _listeners = {};
+    promise = promise || Promise.resolve(null);
+
+    // And this is something else that might be reusable...
+    function Listener(reqId, extraData) {
+      _listeners[extraData.type][reqId] = extraData.cb;
+      this.serialize = function() {
+        return {
+          id: reqId,
+          data: extraData,
+          processAnswer: answer => {
+            if (answer.event) {
+              listenerCb(answer.event, _listeners[extraData.type][reqId]);
+            }
+          }
+        };
+      };
+    }
+
+    function ListenerRemoval(reqId, extraData) {
+      this.serialize = function() {
+        return {
+          id: reqId,
+          data: extraData,
+          processAnswer: answer => debug('Got an invalid answer for: ' + reqId)
+        };
+      };
+    }
+
+    function Dispatcher(reqId, extraData) {
+      this.serialize = function() {
+        return {
+          id: reqId,
+          data: extraData,
+          processAnswer: answer => debug('Got an invalid answer for: ' + reqId)
+        };
+      };
+    }
+
+    this.addEventListener = function(type, cb, useCapture) {
```

Comment:
Hasn't this been implemented already at some other place? Not saying you should change it now! Just something to check at a later point.